Assets/Mirage/Weaver/Processors/PropertySiteProcessor.cs

```diff
@@ -46,31 +56,59 @@ void ProcessInstructionGetterField(Instruction i, FieldReference opField)
             // does it set a field that we replaced?
             if (Getters.TryGetValue(opField, out MethodDefinition replacement))
             {
-                //replace with property
-                i.OpCode = OpCodes.Call;
-                i.Operand = replacement;
+                if (opField.DeclaringType.IsGenericInstance || opField.DeclaringType.HasGenericParameters) // We're calling to a generic class
+                {
+                    FieldReference newField = i.Operand as FieldReference;
+                    GenericInstanceType genericType = (GenericInstanceType)newField.DeclaringType;
+                    i.OpCode = OpCodes.Callvirt;
+                    i.Operand = replacement.MakeHostInstanceGeneric(genericType);
+                }
+                else
+                {
+                    //replace with property
+                    i.OpCode = OpCodes.Call;
+                    i.Operand = replacement;
+                }
             }
         }
 
         Instruction ProcessInstruction(MethodDefinition md, Instruction instr, SequencePoint sequencePoint)
         {
             if (instr.OpCode == OpCodes.Stfld && instr.Operand is FieldReference opFieldst)
             {
+                FieldReference resolved = opFieldst.Resolve();
+                if (resolved == null)
+                {
+                    resolved = opFieldst.Duplicate(opFieldst.DeclaringType.Resolve());
+                }
+
                 // this instruction sets the value of a field. cache the field reference.
-                ProcessInstructionSetterField(instr, opFieldst);
+                ProcessInstructionSetterField(instr, resolved);
             }
 
             if (instr.OpCode == OpCodes.Ldfld && instr.Operand is FieldReference opFieldld)
             {
+                FieldReference resolved = opFieldld.Resolve();
+                if (resolved == null)
+                {
+                    resolved = opFieldld.Duplicate(opFieldld.DeclaringType.Resolve());
+                }
+
                 // this instruction gets the value of a field. cache the field reference.
-                ProcessInstructionGetterField(instr, opFieldld);
+                ProcessInstructionGetterField(instr, resolved);
             }
 
             if (instr.OpCode == OpCodes.Ldflda && instr.Operand is FieldReference opFieldlda)
             {
+                FieldReference resolved = opFieldlda.Resolve();
+                if (resolved == null)
+                {
+                    resolved = opFieldlda.Duplicate(opFieldlda.DeclaringType.Resolve());
```

Comment:
Duplicate is a little misleading.   I would rather we just lookup the field in the decclaring type,  like so:
```suggestion
                    resolved = opFieldlda.DeclaringType.Resolve().GetField(opFieldlda.Name);
```

Instead of "Duplicate" method extension,  add a GetField method extension that finds the field by name.

Reply:
Fixed in cfc9a6f.